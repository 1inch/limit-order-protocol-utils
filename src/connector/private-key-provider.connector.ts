import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';
import {MsgParams, signTypedData_v4} from 'eth-sig-util';

export class PrivateKeyProviderConnector implements ProviderConnector {
    constructor(
        private readonly privateKey: string,
        protected readonly web3Provider: Web3
    ) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        _walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash = ''
    ): Promise<string> {
        const result = signTypedData_v4(Buffer.from(this.privateKey, 'hex'), {
            data: typedData,
            /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
        } as MsgParams<any>);

        return Promise.resolve(result);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
