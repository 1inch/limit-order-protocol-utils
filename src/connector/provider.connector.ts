import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem, AbiOutput} from '../model/abi.model';

export interface AbiDecodeResult {
    [key: string]: unknown,
}

export interface ProviderConnector {
    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string;

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        typedDataHash: string
    ): Promise<string>;

    ethCall(contractAddress: string, callData: string): Promise<string>;

    decodeABIParameter<T>(type: AbiOutput | string, hex: string): T;

    decodeABICallParameters(types: Array<AbiOutput | string>, callData: string): AbiDecodeResult;
}
