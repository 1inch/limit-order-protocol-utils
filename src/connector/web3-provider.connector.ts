import {ProviderConnector} from './provider.connector';
import Web3 from 'web3';
import {EIP712TypedData} from '../model/eip712.model';
import {AbiItem} from '../model/abi.model';
import {AbiItem as Web3AbiItem} from 'web3-utils';

export class Web3ProviderConnector implements ProviderConnector {
    constructor(protected readonly web3Provider: Web3) {}

    contractEncodeABI(
        abi: AbiItem[],
        address: string | null,
        methodName: string,
        methodParams: unknown[]
    ): string {
        const contract = new this.web3Provider.eth.Contract(
            abi as Web3AbiItem[],
            address === null ? undefined : address
        );

        return contract.methods[methodName](...methodParams).encodeABI();
    }

    signTypedData(
        walletAddress: string,
        typedData: EIP712TypedData,
        /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
        _typedDataHash: string
    ): Promise<string> {
        if (!this.web3Provider.currentProvider) {
            throw new Error('Web3 currentProvider is null');
        }

        return this.web3Provider.currentProvider.send('eth_signTypedData_v4', [
            walletAddress,
            JSON.stringify(typedData),
        ]);
    }

    ethCall(contractAddress: string, callData: string): Promise<string> {
        return this.web3Provider.eth.call({
            to: contractAddress,
            data: callData,
        });
    }

    decodeABIParameter<T>(type: string, hex: string): T {
        return this.web3Provider.eth.abi.decodeParameter(type, hex) as T;
    }
}
