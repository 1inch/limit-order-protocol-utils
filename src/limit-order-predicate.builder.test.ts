import Web3 from 'web3';
import {Web3ProviderConnector} from './connector/web3-provider.connector';
import {LimitOrderProtocolFacade} from './limit-order-protocol.facade';
import {
    LimitOrderPredicateBuilder,
    LimitOrderPredicateCallData,
} from './limit-order-predicate.builder';
import {Erc20Facade} from './erc20.facade';
import {
    COMPLEX_PREDICATE_SNAPSHOT,
    SIMPLE_PREDICATE_SNAPSHOT,
} from '../test/predicate-snapshots';

describe('PredicateBuilder - for build limit order predicate', () => {
    const contractAddress = '0xe3456f4ee65e745a44ec3bcb83d0f2529d1b84eb';
    const walletAddress = '0xfb3c7eb936caa12b5a884d612393969a557d4307';
    const WBNB_ADDRESS = '0xbb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c';

    const web3 = new Web3('https://bsc-dataseed.binance.org');
    const providerConnector = new Web3ProviderConnector(web3);

    let facade: LimitOrderProtocolFacade;
    let predicateBuilder: LimitOrderPredicateBuilder;
    let erc20Facade: Erc20Facade;

    function buildComplexPredicate(): {
        valuesInPredicate: string[];
        predicate: LimitOrderPredicateCallData;
    } {
        const balanceOfCallData = erc20Facade.balanceOf(
            WBNB_ADDRESS,
            walletAddress
        );

        const timestampBelow1 = 1619860038;
        const nonce1 = 13344354354;
        const gt1 = 245673600000000000;
        const timestampBelow2 = 1619875555;
        const lt1 = 3453466520000000000;
        const eq1 = 34643434200000000;

        const valuesInPredicate = [
            timestampBelow1.toString(16),
            nonce1.toString(16),
            gt1.toString(16),
            timestampBelow2.toString(16),
            lt1.toString(16),
            eq1.toString(16),
        ];

        const {
            or,
            and,
            timestampBelow,
            nonceEquals,
            gt,
            lt,
            eq,
        } = predicateBuilder;

        const predicate = or(
            and(
                timestampBelow(timestampBelow1),
                nonceEquals(walletAddress, nonce1),
                gt(gt1.toString(), WBNB_ADDRESS, balanceOfCallData)
            ),
            or(
                timestampBelow(timestampBelow2),
                lt(lt1.toString(), WBNB_ADDRESS, balanceOfCallData)
            ),
            eq(eq1.toString(), WBNB_ADDRESS, balanceOfCallData)
        );

        return {valuesInPredicate, predicate};
    }

    beforeEach(() => {
        facade = new LimitOrderProtocolFacade(
            contractAddress,
            providerConnector
        );
        erc20Facade = new Erc20Facade(providerConnector);
        predicateBuilder = new LimitOrderPredicateBuilder(facade);
    });

    it('Simple predicate must includes all values and match the snapshot', () => {
        const nonce = 13;
        const timestampBelow = 1619860038;

        const valuesInPredicate = [
            nonce.toString(16),
            timestampBelow.toString(16),
        ];

        const predicate = predicateBuilder.and(
            predicateBuilder.nonceEquals(walletAddress, nonce),
            predicateBuilder.timestampBelow(timestampBelow)
        );

        const allValuesCheck = valuesInPredicate.every((value) =>
            predicate.includes(value)
        );

        expect(predicate).toBe(SIMPLE_PREDICATE_SNAPSHOT.join(''));
        expect(allValuesCheck).toBe(true);
    });

    it('Complex predicate must includes all values and match the snapshot', () => {
        const {valuesInPredicate, predicate} = buildComplexPredicate();

        const allValuesCheck = valuesInPredicate.every((value) =>
            predicate.includes(value)
        );

        expect(predicate).toBe(COMPLEX_PREDICATE_SNAPSHOT.join(''));
        expect(allValuesCheck).toBe(true);
    });

    it('When predicate includes all values in order then check must be true', () => {
        const {valuesInPredicate, predicate} = buildComplexPredicate();

        const allValuesInProperOrder = valuesInPredicate.every(
            (value, index) => {
                const inOrder =
                    predicate.indexOf(valuesInPredicate[index - 1]) <
                    predicate.indexOf(value);

                return index === 0 ? true : inOrder;
            }
        );

        expect(allValuesInProperOrder).toBe(true);
    });

    it('When predicate includes values not in order then check must be false', () => {
        const {valuesInPredicate, predicate} = buildComplexPredicate();
        const firstElement = valuesInPredicate[0];

        // Swap values places
        valuesInPredicate[0] = valuesInPredicate[1];
        valuesInPredicate[1] = firstElement;

        const allValuesInProperOrder = valuesInPredicate.every(
            (value, index) => {
                const inOrder =
                    predicate.indexOf(valuesInPredicate[index - 1]) <
                    predicate.indexOf(value);

                return index === 0 ? true : inOrder;
            }
        );

        expect(allValuesInProperOrder).toBe(false);
    });
});
