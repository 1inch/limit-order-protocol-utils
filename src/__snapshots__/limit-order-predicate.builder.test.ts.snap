// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredicateBuilder - for build limit order predicate Complex predicate must includes all values and match the snapshot 1`] = `"0x742611450000000000000000000000000000000000000000000004ec000003c800000204000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000004ecbfa7514300000000000000000000000000000000000000000000018c00000068000000240000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000018c63592c2b00000000000000000000000000000000000000000000000000000000608d1a46cf6fc6e3000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d4307000000000000000000000000000000000000000000000000000000031b62b0324f38e2b80000000000000000000000000000000000000000000000000368ced66ab10000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a4bf15fcd8000000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d4307000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007426114500000000000000000000000000000000000000000000000000000148000000240000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000014863592c2b00000000000000000000000000000000000000000000000000000000608d56e3ca4ece220000000000000000000000000000000000000000000000002fed2d80b04df000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a4bf15fcd8000000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d430700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006fe7b0ba000000000000000000000000000000000000000000000000007b1405154a6600000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000a4bf15fcd8000000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c0000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000002470a08231000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d430700000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;

exports[`PredicateBuilder - for build limit order predicate Simple predicate must includes all values and match the snapshot 1`] = `"0xbfa75143000000000000000000000000000000000000000000000000000000680000004400000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000000068cf6fc6e3000000000000000000000000fb3c7eb936caa12b5a884d612393969a557d4307000000000000000000000000000000000000000000000000000000000000000d63592c2b00000000000000000000000000000000000000000000000000000000608d1a46000000000000000000000000000000000000000000000000"`;

exports[`PredicateBuilder - for build limit order predicate Simplyfied predicate must includes all values and match the snapshot 1`] = `"0x2cc2878d0000608d1a4600000000000efb3c7eb936caa12b5a884d612393969a557d4307"`;
